### PR TITLE
Adjust caching logic when a `NodeStateTable` is empty

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/57455

For context, here's the relevant logic in `BatchNode`. What happened was that `sourceTable` would be empty (the new source removed the attribute, so no syntax is returned from this source), but `sourceTable.IsCached` would return `true` which leads us to use the old entries `newTable.TryUseCachedEntries(stopwatch.Elapsed, sourceInputs)`.

```
        public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.Builder builder, NodeStateTable<ImmutableArray<TInput>> previousTable, CancellationToken cancellationToken)
        {
            // grab the source inputs
            var sourceTable = builder.GetLatestStateTableForNode(_sourceNode);

            // Semantics of a batch transform:
            // Batches will always exist (a batch of the empty table is still [])
            // There is only ever one input, the batch of the upstream table
            // - Output is cached when upstream is all cached
            // - Added when the previous table was empty
            // - Modified otherwise

            // update the table
            var newTable = builder.CreateTableBuilder(previousTable, _name);

            // If this execution is tracking steps, then the source table should have also tracked steps or be the empty table.
            Debug.Assert(!newTable.TrackIncrementalSteps || (sourceTable.HasTrackedSteps || sourceTable.IsEmpty));

            var stopwatch = SharedStopwatch.StartNew();

            var batchedSourceEntries = sourceTable.Batch();
            var sourceValues = batchedSourceEntries.SelectAsArray(sourceEntry => sourceEntry.State != EntryState.Removed, sourceEntry => sourceEntry.Item);
            var sourceInputs = newTable.TrackIncrementalSteps ? batchedSourceEntries.SelectAsArray(sourceEntry => (sourceEntry.Step!, sourceEntry.OutputIndex)) : default;

            if (previousTable.IsEmpty)
            {
                newTable.AddEntry(sourceValues, EntryState.Added, stopwatch.Elapsed, sourceInputs, EntryState.Added);
            }
            else if (!sourceTable.IsCached || !newTable.TryUseCachedEntries(stopwatch.Elapsed, sourceInputs))
            {
                if (!newTable.TryModifyEntry(sourceValues, _comparer, stopwatch.Elapsed, sourceInputs, EntryState.Modified))
                {
                    newTable.AddEntry(sourceValues, EntryState.Added, stopwatch.Elapsed, sourceInputs, EntryState.Added);
                }
            }

            return newTable.ToImmutableAndFree();
        }
 ```

### CODE REVIEW

**Comment by jcouv** [comment]
@chsienki @dotnet/roslyn-compiler for review. Thanks

**Review by chsienki — APPROVED** [review]
LGTM. Thanks!

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs [hunk 1]
```diff
@@ -2166,6 +2166,69 @@ class C { }
             Assert.Throws<ArgumentNullException>(() => driver.WithUpdatedParseOptions(null!));
         }
 
+        [Fact, WorkItem(57455, "https://github.com/dotnet/roslyn/issues/57455")]
+        public void RemoveTriggeringSyntaxAndVerifySyntaxTreeConsistentWithCompilation()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs:2170]: >VerifySyntaxTreeConsistentWithCompilation [](http://example.com/codeflow?start=45&length=41)

Are we verifying the syntax tree below, or is this just implicit in the test completing without exceptions?

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs:2170]: We are verifying the syntax tree given to the SG comes from the correct compilation. So I would say "both".

**Review by cston — APPROVED** [review]

**Pull request merged**